isa.c

```diff
@@ -296,4 +312,85 @@ static void exec_B(uint32_t instruction)
+        next_state.regs[rd_id] = data; // casting uint8_t to uint32_t zeros top 3 bytes on its own; done to store byte to LSB of rd_id
+    }
+}
+
+// Multiply and Multiply-Accumulate Instructions
+// MUL and MLA produce the low 32 bits of a multiply.
+// TIL: The results of a signed multiply and an unsigned multiply
+// differ only in the upper 32 bits; the low 32 bits remain the same for both.
+// So we can use the MUL and MLA ops unchanged for signed and unsigned operands.
+static void exec_MUL(uint32_t instruction)
+{
+    if (condition_check(curr_state, get_bits(instruction, 31, 28))) {
+        uint32_t rd_id = get_bits(instruction, 19, 16);
+        uint32_t rs_id = get_bits(instruction, 11, 8);
+        uint32_t rm_id = get_bits(instruction, 3, 0);
+
+        uint32_t result = curr_state.regs[rm_id] * curr_state.regs[rs_id]; // we don't care about overflows; no need to set C(arry) flag
+        next_state.regs[rd_id] = result;
+        
+        if (get_bit(instruction, S_BIT)) {
+            set_bit(&(curr_state.CPSR), CPSR_N, get_bit(result, 31));
+            set_bit(&(curr_state.CPSR), CPSR_Z, ((result) ? 0 : 1));
+        }
+    }
+}
 
+static void exec_MLA(uint32_t instruction)
+{
+    if (condition_check(curr_state, get_bits(instruction, 31, 28))) {
+        uint32_t rd_id = get_bits(instruction, 19, 16);
+        uint32_t rn_id = get_bits(instruction, 15, 12);
+        uint32_t rs_id = get_bits(instruction, 11, 8);
+        uint32_t rm_id = get_bits(instruction, 3, 0);
+
+        uint32_t result = (curr_state.regs[rm_id] * curr_state.regs[rs_id]) + curr_state.regs[rn_id]; // we don't care about overflows; no need to set C(arry) flag
+        next_state.regs[rd_id] = result;
+        
+        if (get_bit(instruction, S_BIT)) {
+            set_bit(&(curr_state.CPSR), CPSR_N, get_bit(result, 31));
+            set_bit(&(curr_state.CPSR), CPSR_Z, ((result) ? 0 : 1));
+        }
+    }
+}
+
+static void exec_MOV(uint32_t instruction)
+{
+    if (condition_check(curr_state, get_bits(instruction, 31, 28))) {
+        struct ShifterOperand * shiftop;
+        shiftop = shifter_operand(curr_state, instruction);
+        uint32_t val = shiftop->shifter_operand;
+        uint32_t rd_id = get_bits(instruction, 15, 12);
+        curr_state.regs[rd_id] = val;
+
+        if (get_bit(instruction, S_BIT)) {
+            set_bit(&(curr_state.CPSR), CPSR_N, get_bit(val, 31));
+            set_bit(&(curr_state.CPSR), CPSR_Z, ((val) ? 0 : 1));
+            set_bit(&(curr_state.CPSR), CPSR_C, shiftop->shifter_carry);
+        }
+    }
+}
+
+static void exec_MVN(uint32_t instruction)
+{
+    if (condition_check(curr_state, get_bits(instruction, 31, 28))) {
+        struct ShifterOperand * shiftop;
+        shiftop = shifter_operand(curr_state, instruction);
+        uint32_t val = ~(shiftop->shifter_operand) & 0xFFFFFFFF; // bitwise negation promotes result to (int); bitwise and-ing done to prevent this
```

Comment:
oh so this wasn't actually necessary here as shifter_operand output is 32bits, but yah still leave as is for clarity